crates/adapters/src/static_compile/deserialize_with_context.rs

```diff
@@ -0,0 +1,783 @@
+                }
+
+                Ok(result)
+            }
+        }
+
+        deserializer.deserialize_seq(VecVisitor::new(context))
+    }
+}
+
+macro_rules! deserialize_tuple {
+    ([$num_fields:expr]($($arg:tt),*)) => {
+        #[allow(unused_variables)]
+        #[allow(dead_code)]
+        #[allow(unused_mut)]
+        impl<'de, C, $($arg),*> DeserializeWithContext<'de, C> for ($($arg),*)
+        where
+            $($arg: DeserializeWithContext<'de, C>),*
+        {
+            fn deserialize_with_context<D>(deserializer: D, context: &'de C) -> Result<Self, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                struct TupleVisitor<'de, C, $($arg),*> {
+                    context: &'de C,
+                    phantom: PhantomData<($($arg),*)>,
+                }
+
+                impl<'de, C, $($arg),*> TupleVisitor<'de, C, $($arg),*> {
+                    fn new(context: &'de C) -> Self {
+                        Self {
+                            context,
+                            phantom: PhantomData,
+                        }
+                    }
+                }
+
+                impl<'de, C, $($arg),*> Visitor<'de> for TupleVisitor<'de, C, $($arg),*>
+                where
+                    $($arg: DeserializeWithContext<'de, C>),*
+                {
+                    type Value = ($($arg),*);
+
+                    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                        write!(formatter, concat!("a tuple of size ", stringify!($num_fields)))
+                    }
+
+                    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+                    where
+                        A: SeqAccess<'de>,
+                    {
+                        let mut cnt: i32 = -1;
+                        Ok(($({cnt+=1; seq
+                            .next_element_seed(<$crate::DeserializationContext<C, $arg>>::new(self.context))?
+                            .ok_or_else(|| {
+                                A::Error::invalid_length(cnt as usize, &self)
+                            })?}),*))
+                    }
+                }
+
+                deserializer.deserialize_tuple($num_fields, TupleVisitor::new(context))
+            }
+        }
+
+
+    }
+}
+
+deserialize_tuple!([0]());
+deserialize_tuple!([2](T0, T1));
+deserialize_tuple!([3](T0, T1, T2));
+deserialize_tuple!([4](T0, T1, T2, T3));
+deserialize_tuple!([5](T0, T1, T2, T3, T4));
+deserialize_tuple!([6](T0, T1, T2, T3, T4, T5));
+deserialize_tuple!([7](T0, T1, T2, T3, T4, T5, T6));
+deserialize_tuple!([8](T0, T1, T2, T3, T4, T5, T6, T7));
+deserialize_tuple!([9](T0, T1, T2, T3, T4, T5, T6, T7, T8));
+deserialize_tuple!([10](T0, T1, T2, T3, T4, T5, T6, T7, T8, T9));
+deserialize_tuple!([11](T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10));
+deserialize_tuple!([12](T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11));
```

Comment:
are these used?
For TupleN the compiler generates this code. We can do this for regular tuples as well.

Reply:
only 2-tuples are used at the moment. Regarding `TupleN`, the compiler generates `Deserialize` implementation for them, which is good enough since we don't deserialize them directly from input data stream.  If we ever want to do that, we'll need to implement `DeserializeWithContext` for them too.